### ISSUE TITLE
verifyBeforeUpdateEmail does not unlink the old email address

### ISSUE THREAD

**Issue by kinex** [firebase/firebase-android-sdk#4505]
### [READ] Step 1: Are you in the right place?

I think so.

### [REQUIRED] Step 2: Describe your environment

  * Firebase Component: Auth
  * Firebase android SDK BOM - 31.0.1 (FlutterFire)

#### Steps to reproduce:

1. Create a new account "_a_@_gmail.com_".
2. Call `verifyBeforeUpdateEmail` to update the account email to "_b_@_gmail.com_".
3. Account's new primary email is now "_b_@_gmail.com_" as expected, but the old email "_a_@_gmail.com_" was added as a linked provider. This causes the following problems:
 - User can still sign in with the old email "_a_@_gmail.com_" and it gets linked to the same account as the new "_b_@_gmail.com_". In other words the user cannot use the old email for creating a new account. This is confusing to user.
 - There are now two email addresses with the same `providerId` (google.com). Is this even a valid state in the Firebase Authentication?
 - It is not possible to delete the old email address using `User.unlink`, because it would delete also the new email address as they both have the same `providerId` (https://github.com/firebase/firebase-android-sdk/issues/4504).

Why `verifyBeforeUpdateEmail` keeps the old email address? I think the expected behavior is that the old email address is removed from the account automatically when user clicks the confirmation link in the email.

**Comment by argzdev** [comment]
Hi @kinex, thanks for reporting. I'm unable to reproduce the behavior you're experiencing.

> Account's new primary email is now "b@gmail.com" as expected, but the old email "a@gmail.com" was added as a linked provider. This causes the following problems:

I don't see the old email "a@gmail.com" being added as a linked provider.

Here's the steps I followed using Firebase BoM version `31.0.1`:
1. Register new account via email/password "a@gmail.com"
2. *Optional*: Verify email with `sendEmailVerification`
3. Use `verifyBeforeUpdateEmail` and change email to "b@gmail.com"
4. Open "b@gmail.com" email and accepting the link 
5. Firebase Console shows that "a@gmail.com" is changed to "b@gmail.com"

Am I missing anything? Perhaps you're unintentionally adding the old email somewhere in your code? If not, could you provide a minimal reproducible example of your issue? It'll help us investigate this further. Thanks!

**Comment by kinex** [comment]
Hi @argzdev, thanks for your quick response. The issue is not visible in the Firebase Console because you can see there only the providers with a different provider ID and in this case both email addresses have the same provider ID. Did you check the linked providers using this API https://firebase.google.com/docs/reference/android/com/google/firebase/auth/FirebaseUser#getProviderData() ? 

I didn't test this by using a new email address with a different provider ID than the old email address, but I would assume it has also similar issue.

I am actually using FlutterFire, but it is mostly only a wrapper to the platform SDKs so I believe the issue is not related to the FlutterFire directly.

**Comment by argzdev** [comment]
Hi @kinex, thanks for the extra details. I tried checking the provider ID for both "a@gmail.com" and "b@gmail.com", but both email addresses have no linked providers between each other. I'm not sure if this is an issue in FlutterFire, and you're right that since it is acting as a wrapper, any underlying issue is from the native SDK, but as of the moment of testing this isn't visible on the native Android side, so there could be a problem with the wrapper? I can't say.

Here are the `providerData`:
`providerData` list for "a@gmail.com":
- email: a@gmail.com providerId: firebase
- email: a@gmail.com providerId: password

`providerData` list for "b@gmail.com":
- email: b@gmail.com providerId: firebase
- email: b@gmail.com providerId: password

I think the best approach here is you could try using our [quickstart sample](https://firebase.google.com/docs/samples) to simulate this behavior and see if you're able to reproduce it, this is so that we have a minimal reproducible example. And if you're able to reproduce this, then you could reach out to the [FlutterFire team](https://github.com/firebase/flutterfire) and see if they can reproduce the same behavior with the minimal repro. You may also reference this issue so we can keep track of the conversation.

Thanks!

**Comment by kinex** [comment]
What is providerId "firebase"? I have seen only these https://github.com/firebase/firebase-ios-sdk/blob/master/FirebaseAuth/Sources/AuthProvider/FIRAuthProvider.m (from iOS SDK).

There is probably one major difference in our test steps. I am testing this with Google accounts (Sign in with Google), so both a@gmail.com and b@gmail.com must be Google accounts (not Email/Password accounts). The provider ID must be "google.com". Sorry this is not very clearly mentioned in my bug report. 

I also verified that if you use email authentication (providerId=password) there is no this issue. So the issue may be related to sign in with Google only.

I checked related FlutterFire code and cannot see anything special there. If you still cannot reproduce this, I could test this with native Android but it requires more time.

**Comment by argzdev** [comment]
Hi @kinex, so that was what was missing! I was able to reproduce the behavior you're experiencing. Thank you for working with me on this, I'll notify our engineers and see if this is an intended behavior (most likely not) or a bug.

NOTES: 

Scenario: 
- `a@gmail.com` changed to `b@gmail.com`

UserInfo in `providerData` for email `b@gmail.com`:
- email: `a@gmail.com` providerId: firebase
- email: `b@gmail.com` providerId: google.com
- email: `a@gmail.com` providerId: google.com

UserInfo in `providerData` for email `a@gmail.com`:
- email: `b@gmail.com` providerId: firebase
- email: `a@gmail.com` providerId: google.com
- email: `b@gmail.com` providerId: google.com

**Comment by kinex** [comment]
Great to hear you managed to reproduce the issue! In my opinion `verifyBeforeUpdateEmail` cannot be used with Google accounts before this gets fixed. Current behavior may cause even a security issue if a developer does not notice that the old email address is not unlinked as expected (ex. a use case where a user updates their email address to transfer their account to another user).

**Comment by argzdev** [comment]
Hi @kinex, thank you for your patience. After carefully discussing this with our engineers, it seems that this behavior is currently working as expected, and changing this behavior might impact other developers that are not expecting this change. That said, I have also informed our engineers regarding the feature request #4504. Currently, we can't promise any timeline for the feature request.

For others that find it useful, please add an emoji thumbs up on the original post of #4504. This will help our engineers prioritize adding this to the roadmap depending on demand.

FWIW, I was able to create a workaround regarding your use case, however this only works for those with only single google providerId. 

1. Change a@gmail.com > b@gmail.com
2. Logout, then login as [b@gmail.com](mailto:b@gmail.com) (This is needed because [a@gmail.com](mailto:a@gmail.com)'s google sign in is no longer valid, and using unlink will not work, and will result to `com.google.firebase.auth.FirebaseAuthInvalidUserException: The user's credential is no longer valid. The user must sign in again.`)
3. Use unlink("google.com"), both [a@gmail.com](mailto:a@gmail.com) and [b@gmail.com](mailto:b@gmail.com) with google provider is removed. [b@gmail.com](mailto:b@gmail.com) with providerId firebase is retained.
4. Link [b@gmail.com](mailto:b@gmail.com) with google provider silently using linkWithCredential

The User UID of [b@gmail.com](mailto:b@gmail.com) will not change since the providerId of firebase is retained.

TLDR: "Unlink all, then link current"
